compiler/qsc_fir/src/ty.rs

```diff
@@ -178,20 +178,77 @@ fn instantiate_arrow_ty<'a>(
     })
 }
 
-impl Display for GenericParam {
+impl Display for TypeParameter {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
         match self {
-            GenericParam::Ty => write!(f, "type"),
-            GenericParam::Functor(min) => write!(f, "functor ({min})"),
+            TypeParameter::Ty { name, bounds } => {
+                write!(f, "type ({name}){bounds}")
+            }
+            TypeParameter::Functor(min) => write!(f, "functor ({min})"),
+        }
+    }
+}
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct ClassConstraints(pub Box<[ClassConstraint]>);
+
+impl std::fmt::Display for ClassConstraints {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if self.0.is_empty() {
+            Ok(())
+        } else {
+            let bounds = self
+                .0
+                .iter()
+                .map(std::string::ToString::to_string)
+                .collect::<Vec<_>>()
+                .join(", ");
+            write!(f, "{bounds}")
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub enum ClassConstraint {
+    #[default]
+    Eq,
+    Add,
+    Exp {
+        power: Ty,
+    },
+    Iterable {
+        item: Ty,
+    },
+
+    /// A class that is not built-in to the compiler.
+    NonNativeClass(Rc<str>),
+    Num,
```

Comment:
As we discussed, it is better if we split the `Num` constraint into `Mul`, `Div`, `Sub`, etc. to stay consistent with `Add` and `Exp` being their own contraints.